### PR TITLE
Prefer in-memory task lookups in TaskTracker.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
CC @pyronicide

### CODE REVIEW

**Comment by pyronicide** [comment]
I just tested this, and it looks *WAY* better:

https://www.dropbox.com/s/h6as8c2wzbxzcc9/Screenshot%202015-04-17%2018.18.41.png?dl=0

**Comment by drexin** [comment]
This is great, but I think we should invalidate the in-memory state on loss of leadership.

File: src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala [hunk 1]
```diff
@@ -188,25 +192,39 @@ class TaskTracker @Inject() (
...
+      }
     }
-    else None
+
+    // Search the non-local storage for the supplied task.
+    def fromPersistentStorage(): Option[MarathonTask] = {
+      val bytes = fetchFromState(taskKey).value
+      if (bytes.length > 0) {
+        val source = new ObjectInputStream(new ByteArrayInputStream(bytes))
+        deserialize(taskKey, source)
```
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala]: We should also update the cached state here after fetching from storage.

**Comment by ConnorDoyle** [comment]
@drexin once I make the suggested change do you think this can be included in `0.8.2`?

**Comment by ConnorDoyle** [comment]
Build failing due to inability to download the sbt pgp plugin.  Perhaps the Typesafe maven repo is having trouble?  Restarted the build...

**Comment by ConnorDoyle** [comment]
Added cache invalidation on loss of leadership.  Fetched values are now properly cached.

**Comment by drexin** [comment]
We should add a test that verifies usage of the cache. Since we want to release today, I can add that. I will merge this after some manual testing. Thanks @ConnorDoyle, this is a great improvement.

**Comment by bobrik** [comment]
I tried to build this branch and got strange results.

1. Run 0.8.2-SNAPSHOT from master (or 0.8.1) with 1 task (healthy).
2. Upgrade to snapshot from this branch, restart whole cluster.
3. Task is not present in mesos ui, marathon claims that it's unhealthy. Moreover, you cannot kill this task from marathon ui to reschedule it.
4. Revert back, see your task rescheduled as it should.

As a side note: api is quite slow. I can push mesos master to 480rps for `/state.json`, but marathon can only give me 15rps for `/v2/apps/whatever/app/v1?embed=apps.tasks` for the same cluster with just 1 app with 1 task. Is there specific issue to track this?

File: src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala [hunk 2]
```diff
@@ -188,25 +194,47 @@ class TaskTracker @Inject() (
...
+    def fromPersistentStorage(): Option[MarathonTask] = {
+      val bytes = fetchFromState(taskKey).value
+      if (bytes.length > 0) {
+        val source = new ObjectInputStream(new ByteArrayInputStream(bytes))
+        val taskOption = deserialize(taskKey, source)
+
+        for (task <- taskOption) {
+          // Update the in-memory storage.
+          val internalApp = apps.getOrElseUpdate(appId, new InternalApp(appId, TrieMap(), false))
+          internalApp.tasks += (taskKey -> task)
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala]: This should be putIfAbsent, in case another thread added this task in the meantime.

**Comment by kolloch** [comment]
Hi @ConnorDoyle ,

don't you need some code in store to invalidate your in-memory version?

**Comment by kolloch** [comment]
fetchApp already writes fetched tasks into memory. It calls your reimplemented fetchTaskByKey which already should have done that. Is that intentional?

In `terminated`, we first remove the in-memory version and then expunge the stored version. If we have a read in-between, I think that we have stale data in-memory again.

We can try to come up with more ideas for race conditions, but I really feel that this should be refactored into non-concurrent code and used inside an actor.

At least, the strategies for managing state in this critical piece of code should be documented and then strictly followed.

File: src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala (hunk 2 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala]: Good call, updating.

**Comment by kolloch** [comment]
I close this for now, since we cannot merge this soon.

I think we have to come up with a new PR which uses a different strategy.

**Comment by air** [comment]
kolloch: that patch is pretty problematic. Maybe it works for a demo but I wouldn’t like to rely on that.

ConnorDoyle: Maybe we can address the remaining issues you raised instead of dropping it.  It may still have some concurrency issues (solvable) but round-tripping to ZK so much really limits scalability.
In other words, in this case I'd prefer local lock contention to remote lock contention plus network latency.

**Comment by kolloch** [comment]
@ConnorDoyle: Why not raise this opinion when I closed this PR?

Other than that: Yes, I agree that reads should go to memory. To make that work, the code has to be changed significantly. Your patch trades performance for correctness. I will not usually agree to make that trade off. I would hope that you understand that.

Please support us in doing things right and not add another hack.

**Comment by kolloch** [comment]
@ConnorDoyle We have planned to solve this issue with an actor-based solution. I will start working on that this or next week. The new solution will serve reads from memory.

Unfortunately, this will not be ready for the demo. One of the easiest ways to fix your PR to make it run good enough, would be to use "synchronized" a lot. This will probably still allow us to perform better and would be suitable for the demo. 

I would not like to include it on "master" or in a production version of Marathon.

**Comment by aquamatthias** [comment]
Hey @ConnorDoyle can you reopen that PR with the points of the comments addressed?
That would be very helpful. 
Thanks!

**Pull request closed without merging**